File: app/src/main/java/com/castelijns/mmdemo/users/UsersAdapter.java
package com.castelijns.mmdemo.users;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.castelijns.mmdemo.R;
import com.castelijns.mmdemo.models.User;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class UsersAdapter extends RecyclerView.Adapter<UsersAdapter.ViewHolder> {

    private LayoutInflater inflater;
    private List<User> dataSet;
    private ItemClickListener itemClickListener;

    static class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.tv_name)
        TextView tvName;

        @BindView(R.id.tv_username)
        TextView tvUsername;

        ViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }

    UsersAdapter(Context context, List<User> dataSet) {
        inflater = LayoutInflater.from(context);
        this.dataSet = dataSet;
    }

    @Override
    public UsersAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.row_user, parent, false);
        return new UsersAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(UsersAdapter.ViewHolder holder, final int position) {
        final User user = dataSet.get(position);
        holder.tvName.setText(user.getName());
        holder.tvUsername.setText(user.getUsername());

        holder.itemView.setOnClickListener(v -> {
            if (itemClickListener != null) {
                itemClickListener.onClick(user);
            }
        });
    }

    @Override
    public int getItemCount() {
        return dataSet.size();
    }

    void setItemClickListener(ItemClickListener listener) {
        itemClickListener = listener;
    }

    interface ItemClickListener {
        void onClick(User user);
    }
}

File: app/src/main/kotlin/com/castelijns/mmdemo/users/UsersAdapter.kt
package com.castelijns.mmdemo.users

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.RecyclerView
import com.castelijns.mmdemo.R
import com.castelijns.mmdemo.models.User
import kotlinx.android.synthetic.main.row_user.view.*

class UsersAdapter(
        private val dataSet: List<User>
) : RecyclerView.Adapter<UsersAdapter.ViewHolder>() {

    private lateinit var inflater: LayoutInflater
    private var itemClickListener: ItemClickListener? = null

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.row_user, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val user = dataSet[position]
        holder.itemView.apply {
            tv_name.text = user.name
            tv_username.text = user.username
            tv_email.text = user.email
        }

        holder.itemView.setOnClickListener {
            if (itemClickListener != null) {
                itemClickListener!!.onClick(user, holder.itemView.tv_username, holder.itemView.tv_email)
            }
        }

        // For shared element transitions.
        ViewCompat.setTransitionName(holder.itemView.tv_username, String.format(
                "%s.%s", "username", user.username))
        ViewCompat.setTransitionName(holder.itemView.tv_email, String.format(
                "%s.%s", "email", user.email))
    }

    override fun getItemCount(): Int {
        return dataSet.size
    }

    fun setItemClickListener(listener: ItemClickListener) {
        itemClickListener = listener
    }

}

interface ItemClickListener {
    fun onClick(user: User, tvUsername: TextView, tvEmail: TextView)
}
